### ISSUE TITLE
UWP: Support arbitrary file locations

### ISSUE THREAD

**Issue by sjb-sjb** [aspnet/Microsoft.Data.Sqlite#287]
I previously suggested that user file locations be supported by SQLite under Entity Framework (see https://github.com/aspnet/EntityFramework.Docs/issues/156#issuecomment-231568985 ):

The issue was closed with the comment that absolute paths are already supported, and if the problem still exists in Entity Framework Core 1.0.0 then please open another issue here. 

Well, the problem still exists in EF Core 1.0.0: you can specify an absolute path but only if it is in the app dir. I want to specify a path in the user's document directory, in order to use the SQLite file as a user document format. 

Being able to store the file in arbitrary locations is important to apps where the file is being used as a document format or where it is otherwise desired to allow the user to move, copy or backup the file, or even just because the user wants to be able to see it in with their other files. 

I know the file should be on a local drive but I think we can do better than having it hidden away in a directory that most users can't find. 

To reproduce the problem, download and build the UWP.SQLite example from the Entity Framework Core git site, then change "Filename=Blogging.db" to @"Filename=C:\Users<username>\Blogging.db" in OnConfiguring. Run it and see SQLite error 14: cannot open database file.

**Comment by bricelam** [comment]
UWP apps are sandbox and their [file access permissions](https://msdn.microsoft.com/en-us/windows/uwp/files/file-access-permissions) are extremely restricted.

The native SQLite library calls the Win32 for UWP apps [CreateFile2](https://msdn.microsoft.com/en-us/library/windows/apps/hh449422.aspx) API which, according to the documentation:

> ...can open only files or directories inside the ApplicationData.LocalFolder or Package.InstalledLocation directories.

So in layman's terms: It can't be done.

One alternative is to allow users to import/export their database file from/to arbitrary locations [using a picker](https://msdn.microsoft.com/en-us/windows/uwp/files/quickstart-using-file-and-folder-pickers).

**Comment by fgoulet** [comment]
It can be done by using a Virtual File System as described in this article [Using-SQLite-Databases-in-Any-Folder-in-WinRT](http://www.codeproject.com/Articles/879846/Using-SQLite-Databases-in-Any-Folder-in-WinRT). On GitHub: [SQLiteWinRTVFS](https://github.com/peter0302/SQLiteWinRTVFS).

I have successfully used the VFS with [sqlite-winrt](https://sqlwinrt.codeplex.com/) but not with Microsoft.Data.Sqlite. An exception is thrown in the method GetStorageFileFromPath and I was not  able to find the cause yet. But I am pretty sure there is no technical reason that would prevent it to work eventually.

I would love to see SQLiteWinRTVFS  integrated in Microsoft.Data.Sqlite repository. This would greatly widening the usage of sqlite on the UWP platform.

**Comment by bricelam** [comment]
Oh sweet, I didn't know a VFS existed for that. I thought it was probably possible, but wasn't about to recommend anyone write it themself. 😉

**Comment by divega** [comment]
**EF Triage:** we would welcome a member of the community to contribute a fix to unblock using our provider with the VFS extension mentioned above, but we are not planning to work on it in the near future.

**Comment by sjb-sjb** [comment]
Well, the approach I've taken for the time being is to copy the file back and forth. There is a certain cheezyness to this approach but it does have advantages if the file is located on a network drive. 

The interesting part, though, was getting the dependency injection to relate to the file manager. I decided to use Microsoft.Extensions.DependencyInjection and have it enter a scope when the database is "opened" (i.e. copied into the appdir for use) and exit the scope when the database is "closed" (i.e. copied out of the appdir to the user location). It fits together nicely, you can have services that are available just when the file is open. I am using this in a UWP application. 

One feature of the MS DI that it brought to light is the fact that you can create 'scoped' service objects even if there is actually no scope active ... the created service object in that case is actually a permanent singleton. Fortunately it is easy to avoid this weird behavior (assuming you know about it). 

On the EF side, what I noticed is that the creation and use of a DbContext is not coupled with file management. Even if you do have arbitrary file locations, you don't have a very simple relationship between a given DbContext and a given file. There are document-oriented scenarios where you want to be able to use multiple files at a time (all of the same type) each connected to a different set of DbContext objects. It can be done with the current initialization design of DbContext, but at the cost of writing more code -- if you are going to support file locations you might want to think about making multiple-file scenarios easy. 

- sjb

**Comment by bricelam** [comment]
Closing this as "won't fix", but the triage decision above still stands--if anyone wants to contribute improvements to our provider to enable this, we'd take them.

**Comment by sjb-sjb** [comment]
I would like to ask that this issue be re-opened and reconsidered. 

The official philosophy of app data vs user data includes "Don't use app data to store user data or anything that users might perceive as valuable and irreplaceable." and specifically "User data includes [...] database records holding content created by the user." In contrast, app data is for "runtime state, user preferences, and other settings" and is "ideal for storing app-specific user preferences, settings, and favorites". The distinction is that app data will be blown away whenever the app is deleted or re-installed, which can easily happen. This is all from you (Microsoft) at https://docs.microsoft.com/en-us/windows/uwp/design/app-settings/store-and-retrieve-app-data 

So, if EF with SQLite is restricted to the app dir then it can only store app data (relatively unimportant configuration data), not "real" data that users actually want. It seems unacceptable for something that is MS's preferred data architecture. 

The approach of copying files back and forth is not adequate. Aside from the inefficiency, it becomes more or less impossible to implement file properties (see https://stackoverflow.com/questions/47842474/saving-system-comment-or-system-subject-file-properties-in-uwp/47899048#47899048). 

I also have little confidence in the idea of a virtual file system to solve this problem. It seems like overkill to say the least, and it also doesn't work at this point according to the comments above. 

This is not really a file permissions problem. For example, permissions on the user file can easily be obtained through a file picker.  

Based on @bricelam 's comment above, perhaps the problem is rooted in the fact that SQLite uses Win32 for the UWP CreateFile2 API ?  Perhaps the SQLite file create should be retargeted. This does not seem insurmountable at all. 

Microsoft's EF project should work with the SQLite project to get this fixed.

**Comment by bricelam** [comment]
@ericsink Have you heard of any community activity in this area?

**Comment by ericsink** [comment]
No, I have not.  Interesting thoughts here, though.

**Comment by bricelam** [comment]
I wonder if the new [broadFileSystemAccess](https://docs.microsoft.com/en-us/windows/uwp/files/file-access-permissions#accessing-additional-locations) capability will make this Just Work™.

**Comment by rakismelange** [comment]
@bricelam Any path forward? This is stopping my and other developers adoption of UWP.

broadFileSystemAccess capability Does Not Work™. 
as seen https://github.com/aspnet/Microsoft.Data.Sqlite/issues/543

Is there some other group we can raise the broadFileSystemAccess issue to?

Thanks in advance,

**Comment by tedrog36** [comment]
@bricelam @ericsink  This is also creating problems for my UWP app also.  It feels like setting broadFileSystemAccess is taking a sledge hammer to the problem and it doesn't work now.  It would be great if this were dealt with at the lower-level sqlite layers.

**Comment by bricelam** [comment]
Probably worth starting a discussion on [sqlite-users](http://mailinglists.sqlite.org/cgi-bin/mailman/listinfo/sqlite-users).
